### DISCUSSION TITLE
opening multiple windows for ssh into multiple nodes a la xpanes

### DISCUSSION THREAD

**Discussion by grimm26** [kovidgoyal/kitty#5108] [Q&A] [answered]
I'm a former tmux user and I used xpanes when I needed to open multiple panes in the same tmux window, each ssh'd into a remote node, and broadcast commands to them all.

I'm trying to do something similar with kitty and I came up with this shell function so far:
```
kpanes() {
  local now=$(date +'%s')
  local socket="unix:/tmp/kitmux-${now}"

  kitty -o allow_remote_control=yes --listen-on $socket &
  sleep 1
  kitty @ --to $socket goto-layout grid
  for host in $@; do
    local win_id=$(kitty @ --to $socket new-window --title $host)
    kitty @ --to $socket kitten --match id:${win_id} ssh $host
  done
  kitty @ --to $socket kitten --match id:1 broadcast
}
```
Usage: kpanes host1 host2 host3 ...

I found that if I did not put the sleep that the socket file would not be there and ready for the next kitty commands. Is there a way to wait for the socket to be ready or to have the initial kitty command automagically background itself when ready and allow the rest of the script to carry on? Maybe I would have to make a kitten for this instead of trying to script it.

**Comment by kovidgoyal** [comment]
Just use the broadcast kitten.

**Reply by grimm26** [reply]
You'll notice that I **am** using the broadcast kitten. First I open all the windows, then I start broadcasting.

**Comment by kovidgoyal** [comment, marked as answer]
Ah your question is about how to open multiple kitty windows at startup? Use kitty sessions for that. You can dynamically generate the session file. If you want to use remote control then just wait for the socket to be ready, any number of tools like socat/netcat and similar can do that.

**Reply by grimm26** [reply]
Ah, session file seems simpler. Starting with this:
```
kpanes() {
  local now=$(date +'%s')
  local session_file="/tmp/kpanes-session.${now}"

  # generate session
  echo "layout grid" > $session_file
  for host in $@; do
    echo "launch --title $host kitty +kitten ssh $host" >> $session_file
  done
  echo "launch --allow-remote-control kitty +kitten broadcast" >> $session_file
  kitty --session $session_file
  rm $session_file
}
```

Works more smoothly. Now I just need to do a split stack instead to make the broadcast window at the bottom and the full width but only a few lines tall. Thanks.

**Reply by ApatheticSheep** [reply]
> Now I just need to do a split stack instead to make the broadcast window at the bottom and the full width but only a few lines tall. 

How'd you manage to accomplish this?
I currently have this for syncing keystrokes between kitty windows.
`map ctrl+k launch --allow-remote-control kitty +kitten broadcast --match-tab state:focused`
But the window is big, in the way and annoying.

**Reply by grimm26** [reply]
yes it is annoying. I have just dealt with it because I don't do this very often.
